### PR TITLE
Create a filestorage interface and add support for AWS S3

### PR DESCRIPTION
**Pull request by ftsamis**
Processor stores files in the filesystem. With this commit it becomes possible to choose through the config if the files will be stored on a local filesystem or on an AWS S3 bucket.

Additionally, storing a file to a job specified S3 bucket is supported using the `s3_bucket` and `s3_region` JSON keys. The job submitter needs to have taken care of the downloader's IAM access to the bucket. In this case, requesting a callback (through `callback_dst` or `callback_url`) becomes optional, since the requester's S3 bucket can be used as an event trigger.

Example JSON payload (partial):
```
'{... "s3_bucket": "testbucket", "s3_region": "eu-west-2"}'
```

A filestorage sub-section is introduced under the processor config section.

AWS access is expected to be configured through env vars or an AWS config file.

When a filestorage section is defined, `storage_dir` becomes the temporary storage location. Files under this path are deleted after sent to the filestorage backend.

**config.json examples**

Filesystem backend:
```
"processor": {
"filestorage": {
	"type": "filesystem",
	"rootdir": "/var/downloader_data",
},
"storage_dir":"/tmp/",
```

AWS S3 backend:
```
"processor": {
"filestorage": {
	"type": "s3",
        "region": "eu-west-2",
	"bucket": "downloader",
},
"storage_dir":"/tmp/",
```

### CODE REVIEW

File: processor/filestorage/fileystem.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	RootDir string
+}
+
+func NewFileSystem(rootdir string) *FileSystem {
+	_ = os.MkdirAll(rootdir, os.FileMode(0755))
+	return &FileSystem{RootDir: rootdir}
+}
+
+func (fs FileSystem) StoreFile(srcpath string, destpath string) derrors.DownloadError {
+	fulldestpath := path.Join(fs.RootDir, destpath)
```
Comment (jfragoulis) [inline processor/filestorage/fileystem.go:20]: camel case `fullDestPath` `destPath` `srcPath` (applies to all methods).

File: processor/filestorage/fileystem.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	err := os.MkdirAll(filepath.Dir(fulldestpath), os.FileMode(0755))
+	if err != nil {
+		return derrors.E("creating download directory", err).Internal().Retriable()
+	}
+
+	err = os.Rename(srcpath, fulldestpath)
+	if err != nil {
+		return derrors.E("moving file to perm location", err).Internal().Retriable()
+	}
+	return nil
```
Comment (jfragoulis) [inline processor/filestorage/fileystem.go:30]: Leave an empty line above just for fun (applies to all methods).

File: processor/filestorage/fileystem.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+// DeleteFile removes a file from the filesystem storage
+func (fs FileSystem) DeleteFile(filepath string) derrors.DownloadError {
+	abspath := path.Join(fs.RootDir, filepath)
+	err := os.Remove(abspath)
+	return derrors.E("deleting file", err).Internal().Retriable()
```
Comment (jfragoulis) [inline processor/filestorage/fileystem.go:37]: I guess this returns `nil` if `err` is `nil` ?

File: processor/filestorage/s3.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+func NewAWSS3(bucket string) *AWSS3 {
+	s3Session, err := session.NewSession()
+	if err != nil {
+		// FIXME
+		return &AWSS3{"", nil, nil} //derrors.E("creating AWS S3 session", err)
+	}
+
+	return &AWSS3{bucket: bucket,
```
Comment (jfragoulis) [inline processor/filestorage/s3.go:25]:
```
return &AWSS3{
  bucket: bucket,
```

File: processor/filestorage/s3.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	bucket   string
+	uploader *s3manager.Uploader
+	S3Client *s3.S3
+}
+
+func NewAWSS3(bucket string) *AWSS3 {
+	s3Session, err := session.NewSession()
+	if err != nil {
+		// FIXME
+		return &AWSS3{"", nil, nil} //derrors.E("creating AWS S3 session", err)
```
Comment (jfragoulis) [inline processor/filestorage/s3.go:22]: I guess this should return `nil` if we cant create a session and the caller should handle the error itself.

File: processor/filestorage/s3.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+		Bucket: aws.String(b.bucket),
+		Key:    aws.String(destpath),
+		Body:   f,
+	})
+	if err != nil {
+		return derrors.E("uploading file to S3", err).Internal().Retriable()
+	}
+	os.Remove(srcpath)
+	return nil
+}
```
Comment (jfragoulis) [inline processor/filestorage/s3.go:48]: Εχεις τσακωθεί με τις κενές γραμμες;;;

**Review by jfragoulis — COMMENTED** [review]
Good work!

**Comment by dblia** [comment]
Thanks @ftsamis for the filestorage interface and the AWS S3 storage support!

Regarding S3 we'll also need to provide the ability of using separate S3 bucket, either per aggregation or via an http param, we may debate on this, since we'll want to provide different lifecycle rules and settings per application's bucket. However, I don't believe that we should block the current PR, let's proceed with merging and deploying it in order to ensure that everything's working as expected and we can handle the multiple buckets support on an another PR.

**Comment by ftsamis** [comment]
@dblia Hi, the custom S3 bucket (per job) logic has been already implemented. Updating the description to match the relevant code.

**Review by dblia — APPROVED** [review]
LGTM

**Pull request merged**